### PR TITLE
Implement Counters

### PR DESCRIPTION
**Pull request by ejj**
A simple event counter can be an incredibly helpful debugging tool.
    They're cheap enough to always be running, even in production.
    They're unobtrusive to the average user.  Yet, they can convey a bunch
    of information about the behavior of the system that would otherwise
    be completely opaque.

    This patch implements a new package, counters for Quilt.  It also
    implements a new command, `quilt counters` which can query counters
    from the daemon (querying from the minions is not yet implemented).

### CODE REVIEW

**Comment by kklin** [comment]
I'll review the counters patch because I'll be hacking on it after it's merged anyways

**Comment by kklin** [comment]
@ejj your branch has merge conflicts right now, can you rebase before I review?

**Comment by ejj** [comment]
@kklin rebased

File: minion/ovsdb/ovsdb.go [hunk 1]
```diff
@@ -233,6 +242,7 @@ func parseLogicalSwitchPort(port row) (SwitchPort, error) {
 
 // CreateSwitchPort creates a new logical port in OVN.
 func (ovsdb client) CreateSwitchPort(lswitch string, lport SwitchPort) error {
+	c.Inc("CreateSwitchPort")
```
Comment (kklin) [inline minion/ovsdb/ovsdb.go]: Spaces?

File: api/pb/pb.proto [hunk 2]
```diff
@@ -18,11 +19,23 @@ message DeployRequest {
     string Deployment = 1;
 }
 
-message DeployReply {
-}
+message DeployReply {}
 
 message VersionRequest {}
 
+message CountersRequest {}
```
Comment (ejj) [inline api/pb/pb.proto]: 👍

File: counter/counter.go [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+// Dump returns a list of all in no particular order.
+func Dump() []*pb.Counter {
+	var result []*pb.Counter
+	dumpMutex.Lock()
+	all.Range(func(key, value interface{}) bool {
+		counter := value.(*pb.Counter)
+		cpy := *counter
+
+		// This is not thread-safe if any other code in this module can update
+		// PrevValue.  It's only thread-safe here because of of the dumpMutex.
```
Comment (ejj) [inline counter/counter.go]: 👍

File: quiltctl/command/counter.go [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+	"sort"
+	"text/tabwriter"
+
+	"github.com/quilt/quilt/api/pb"
+)
+
+var usage = `usage: quilt counters [-H=<daemon_host>]
+quilt counters displays internal counters tracked for
+debugging purposes.  It's intended for Quilt experts.
+
```
Comment (ejj) [inline quiltctl/command/counter.go]: Yes, the output looks weird without it.

File: api/client/client.go [hunk 5]
```diff
@@ -48,6 +48,9 @@ type Client interface {
 	// QueryClusters retrieves cluster information tracked by the Quilt daemon.
```
Comment (ejj) [inline api/client/client.go]: 👍

File: quiltctl/command/counter_test.go [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+
+	counters := &Counters{}
+
+	flags := &flag.FlagSet{}
+	assert.Nil(t, flags.Usage)
+
+	counters.InstallFlags(flags)
+
+	assert.NotNil(t, flags.Usage)
+
```
Comment (ejj) [inline quiltctl/command/counter_test.go]: 👍

File: quiltctl/command/counter_test.go [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+}
+
+func TestParse(t *testing.T) {
+	counters := &Counters{}
```
Comment (ejj) [inline quiltctl/command/counter_test.go]: 👍

File: quiltctl/command/counter_test.go [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+	assert.Nil(t, counters.Parse(nil))
+}
+
+func TestRun(t *testing.T) {
+	counters := &Counters{}
```
Comment (ejj) [inline quiltctl/command/counter_test.go]: 👍

File: quiltctl/command/counter_test.go [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+
+	mock.On("QueryCounters").Once().Return(nil, assert.AnError)
+	assert.NotZero(t, counters.Run())
+
+	mock.On("QueryCounters").Once().Return(nil, nil)
+	assert.Zero(t, counters.Run())
+}
+
+func TestPrintCounters(t *testing.T) {
+	counters := []pb.Counter{{
```
Comment (ejj) [inline quiltctl/command/counter_test.go]: 👍

File: minion/network/nat.go [hunk 10]
```diff
@@ -283,6 +289,7 @@ func setDefaultRules(ipt IPTables) error {
 		},
 	}
 	for _, r := range rules {
+		iptC.Inc("Append")
```
Comment (kklin) [inline minion/network/nat.go]: This should be `AppendUnique` to avoid polluting the other `Append` in this package

File: quiltctl/quiltctl.go [hunk 11]
```diff
@@ -20,6 +20,7 @@ var commands = map[string]command.SubCommand{
 	"stop":       command.NewStopCommand(),
 	"version":    command.NewVersionCommand(),
 	"debug-logs": command.NewDebugCommand(),
+	"counters":   &command.Counters{},
```
Comment (ejj) [inline quiltctl/quiltctl.go]: 👍

File: cluster/amazon/client/client.go [hunk 12]
```diff
@@ -6,6 +6,7 @@ import (
 	"github.com/aws/aws-sdk-go/aws"
```
Comment (ejj) [inline cluster/amazon/client/client.go]: I'm not sure I have a good principle on this ...  I'd prefer to wait to add something until we've have some sense of what works and doesn't.

File: cluster/foreman/foreman.go [hunk 13]
```diff
@@ -198,8 +213,10 @@ func updateConfig(m *minion) {
 }
 
 func newClientImpl(ip string) (client, error) {
+	c.Inc("New Minion Client")
 	cc, err := grpc.Dial(ip+":9999", grpc.WithInsecure())
 	if err != nil {
+		c.Inc("New Minion Client Error")
```
Comment (ejj) [inline cluster/foreman/foreman.go]: In this case I did it because we don't log the error anywhere.

File: minion/etcd/connection.go [hunk 14]
```diff
@@ -30,12 +30,14 @@ func runConnectionOnce(conn db.Conn, store Store) error {
 	}
 
 	if conn.EtcdLeader() {
+		c.Inc("Run Connetion Leader")
```
Comment (ejj) [inline minion/etcd/connection.go]: 👍

File: minion/etcd/container.go [hunk 15]
```diff
@@ -31,9 +31,11 @@ func runContainerOnce(conn db.Conn, store Store) error {
 	}
 
 	if conn.EtcdLeader() {
+		c.Inc("Run Container Leader")
 		return updateLeader(conn, store, etcdStr)
 	}
 
+	c.Inc("Run Container Non-Leader")
```
Comment (ejj) [inline minion/etcd/container.go]: 👍

File: minion/ovsdb/ovsdb.go (hunk 1 above)
Comment (ejj) [inline minion/ovsdb/ovsdb.go]: 👍

**Comment by ejj** [comment]
Updated, will merge when travis is happy.

**Pull request merged**